server/src/main/java/org/elasticsearch/index/mapper/SourceLoader.java

```diff
@@ -185,6 +185,10 @@ public Source source(LeafStoredFieldLoader storedFieldLoader, int docId) throws
 
             @Override
             public void write(LeafStoredFieldLoader storedFieldLoader, int docId, XContentBuilder b) throws IOException {
+                for (var fieldLevelStoredFieldLoader : storedFieldLoaders.values()) {
+                    fieldLevelStoredFieldLoader.advanceToDoc(docId);
+                }
```

Comment:
Should this be moved to the loop below, right before calling `load`? I wonder if we end up advancing loaders that are not needed..

Reply:
The loop below is executed only if there are values for the field, so if we do this there we have the same problem with stale values.